asm-lsp-cfg/src/main.rs

```diff
@@ -0,0 +1,434 @@
+    let mut display_entries = Vec::new();
+    let mut path_entries = Vec::new();
+    loop {
+        let selection_text = format!("{}", true_path.display());
+        // Dummy entry to account for the accept option as the first displayed
+        // option
+        path_entries.push(PathBuf::new());
+        display_entries.push("<Select This Directory>".to_string());
+        let Ok(dir_reader) = std::fs::read_dir(&true_path) else {
+            fallback_enter(&mut true_path);
+            return true_path;
+        };
+        for entry in dir_reader.filter_map(std::result::Result::ok) {
+            let entry_path = entry.path();
+            path_entries.push(entry_path.clone());
+            if let Some(name) = entry_path.file_name() {
+                display_entries.push(name.to_string_lossy().to_string());
+            }
+        }
+        let path_selection = FuzzySelect::with_theme(&ColorfulTheme::default())
+            .with_prompt(&selection_text)
+            .default(0)
+            .items(&display_entries[..])
+            .interact()
+            .unwrap();
+
+        // Select current value of `true_path`
+        if path_selection == 0 {
+            break;
+        }
+
+        true_path.clone_from(&path_entries[path_selection]);
+        if true_path.is_file() {
+            break;
+        }
+        path_entries.clear();
+        display_entries.clear();
+    }
+
+    true_path
+}
+
+fn prompt_project(opts: &GenerateOpts) -> ProjectConfig {
+    let path = prompt_project_path(opts);
+    let config = prompt_config();
+
+    ProjectConfig { path, config }
+}
+
+fn prompt_compiler() -> Option<String> {
+    if !Confirm::with_theme(&ColorfulTheme::default())
+        .with_prompt("Provide compiler to generate diagnostics?\n(Used with `compile_flags.txt` files and the following (optional) compile flags field)")
+        .default(true)
+        .interact()
+        .unwrap() {
+        return None;
+    }
+    let mut comp: String;
+    loop {
+        comp = Input::with_theme(&ColorfulTheme::default())
+            .with_prompt("Enter Compiler")
+            .interact_text()
+            .unwrap();
+
+        // Attempt to provide some soft validation, warn the user if something
+        // looks fishy
+        let looks_ok = if comp.contains('/') {
+            // Treat it as a path, try check if it exists
+            let path = PathBuf::from(&comp);
+            let exists = path.exists();
+            let is_file = path.is_file();
+            if !exists {
+                println!(
+                    "Warning: File does not exist at path \"{}\"",
+                    path.display()
+                );
+            } else if !is_file {
+                println!(
+                    "Warning: Path \"{}\" does not point to a file",
+                    path.display()
+                );
+            }
+            !(!exists || !is_file)
+        } else {
+            let find_cmd = if cfg!(target_os = "windows") {
```

Comment:
I had originally wrote code to traverse over all the directories in $PATH here to search for the inputted compiler name. It took several seconds of an invalid name was given, so I opted for this approach instead. If it can be optimized instead, I'd happily replace this hack.

Relevant code snippet:

```rust
let mut found = false;
'path_var_loop: for entry in path_var.split(':').filter(|e| !e.is_empty()) {
    let dir_path = PathBuf::from(entry);
    if let Ok(dir_reader) = std::fs::read_dir(&dir_path) {
        if dir_reader
            .filter_map(std::result::Result::ok)
            .filter(|entry| entry.path().is_file())
            .map(|file| file.file_name())
            .find(|name| name.eq(comp.as_str()))
            .is_some_and(|_| true)
        {
            found = true;
            break 'path_var_loop;
        }
    }
}

if !found {
    println!("Warning: Unable to find provided compiler on $PATH");
}
```

Reply:
```rust
pub fn is_executable(cmd: &str) -> bool {
    use std::env;
    use std::fs;
    use std::path::Path;
    // Check if a path points to an executable file
    let is_executable = {
        |path: &Path| -> bool {
            if path.is_file() {
                #[cfg(unix)]
                {
                    // On Unix, check the `x` bit
                    use std::os::unix::fs::PermissionsExt;
                    let metadata = fs::metadata(path).unwrap();
                    metadata.permissions().mode() & 0o111 != 0
                }
                #[cfg(windows)]
                {
                    // On Windows, check for common executable extensions
                    let extensions = ["exe", "cmd", "bat", "com"];
                    if let Some(ext) = path.extension().and_then(|s| s.to_str()) {
                        return extensions.contains(&ext);
                    }
                    false
                }
            } else {
                false
            }
        }
    };

    // Get the PATH environment variable
    let path_var = env::var_os("PATH").unwrap();

    for path in env::split_paths(&path_var) {
        let full_path = path.join(cmd);
        if is_executable(&full_path) {
            return true;
        }
    }
    false
}
```

This might help, try it out

Reply:
This is a huge improvement, thank you!